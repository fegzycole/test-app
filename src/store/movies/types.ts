export const SET_MOVIES = 'SET_MOVIES';
export const SET_ERROR = 'SET_ERROR';

export interface MovieData {
    imdbID: string;
    Title: string;
    Year: string;
    Type: string;
    Poster: string;
}

export interface MoviesState {
    movies?: MovieData[];
    fetchingMovies: boolean;
    error: string;
}

export interface SetMovies {
    type: typeof SET_MOVIES;
    payload: MovieData[];
}

export interface SetError {
    type: typeof SET_ERROR;
    payload: string;
}

export type MoviesActionTypes = SetMovies | SetError;
