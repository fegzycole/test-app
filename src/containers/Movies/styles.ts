import Grid from '@mui/material/Grid';
import { styled } from '@mui/system';

export const StyledGrid = styled(Grid, {
    name: 'StyledGrid',
})`
    margin: 0 auto;
`;

export const Container = styled(Grid, {
    name: 'Container',
    slot: 'Wrapper',
})`
    width: 90%;
    margin: 40px auto;
    padding: 40px;
`;
