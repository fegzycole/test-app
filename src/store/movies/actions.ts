import { Dispatch } from 'react';
import { MovieData, SetMovies, SET_MOVIES, SetError, SET_ERROR } from './types';
import axios from 'axios';

const apiKey = process.env.REACT_APP_API_KEY;
const apiUrl = process.env.REACT_APP_API_URL;

export const setMovies = (payload: MovieData[]): SetMovies => {
    return {
        type: SET_MOVIES,
        payload,
    };
};

export const setError = (payload: string): SetError => {
    return {
        type: SET_ERROR,
        payload,
    };
};

export const getMovies = (
    searchText: string,
): ((dispatch: Dispatch<{ type: string; payload?: unknown }>) => Promise<void>) => {
    return async (dispatch) => {
        dispatch(setError(''));

        try {
            const request = await axios.get(`${apiUrl}?s=${searchText}&apikey=${apiKey}`);

            const responseData = request.data.Search;

            const movies: MovieData[] = responseData ? (responseData as MovieData[]) : [];

            dispatch(setMovies(movies));
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
            dispatch(setError('An error occurred, check your internet and try again'));
        }
    };
};
