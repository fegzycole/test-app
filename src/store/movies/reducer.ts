import { MoviesActionTypes, MoviesState, SET_MOVIES, SET_ERROR } from './types';

const initialState: MoviesState = {
    movies: undefined,
    fetchingMovies: false,
    error: '',
};

const moviesReducer = (state: MoviesState = initialState, action: MoviesActionTypes) => {
    switch (action.type) {
        case SET_MOVIES:
            return {
                ...state,
                movies: action.payload,
            };
        case SET_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        default:
            return state;
    }
};

export default moviesReducer;
