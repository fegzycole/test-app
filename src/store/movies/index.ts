import * as Types from './types';
import moviesReducer from './reducer';
import * as Actions from './actions';

export { Types, moviesReducer, Actions };
