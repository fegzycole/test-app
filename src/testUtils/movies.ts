import { MovieData } from '../store/movies/types';

export const mockMovie: MovieData = {
    Title: 'Mock Title',
    Type: 'Mock Type',
    imdbID: 'Mock ID',
    Year: '2022',
    Poster: 'Mock Poster',
};

export const mockMovies: MovieData[] = [mockMovie];
