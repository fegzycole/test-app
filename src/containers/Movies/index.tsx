import React, { useState } from 'react';
import SearchBar from '../../components/SearchBar';
import { useSelector } from 'react-redux';
import { RootState } from '../../store';
import { StyledGrid, Container } from './styles';
import MessageBox from '../../components/MessageBox';
import MovieCard from '../../components/MovieCard';

export const Movies: React.FC = () => {
    const { movies } = useSelector((state: RootState) => state.movies);
    const [searchText, setSearchText] = useState<string>('');

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setSearchText(e.target.value);
    };

    const getDisplayedMovies = () => {
        if (!movies) {
            return (
                <MessageBox text="Welcome to Our service, please type the title of the movie you would like to search for in the input field above!" />
            );
        }

        if (!movies.length) {
            return <MessageBox text={`No Movie(s) with the title "${searchText}" found`} />;
        }

        return movies.map((movie) => <MovieCard movie={movie} key={movie.imdbID} />);
    };

    return (
        <Container>
            <SearchBar searchText={searchText} handleChange={handleChange} />
            <StyledGrid container spacing={4}>
                {getDisplayedMovies()}
            </StyledGrid>
        </Container>
    );
};

export default Movies;
