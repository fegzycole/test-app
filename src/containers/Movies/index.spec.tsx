import { cleanup, render } from '@testing-library/react';
import { useSelector } from 'react-redux';
import { mockMovies } from '../../testUtils/movies';
import Movies from './index';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const mockUseSelector = useSelector as jest.Mock<any>;

describe('<Movies />', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });

    describe('when movies is undefined', () => {
        beforeEach(() => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockUseSelector.mockImplementation((selectorFn: any) =>
                selectorFn({
                    movies: {
                        movies: undefined,
                    },
                }),
            );
        });

        it('should render correctly', () => {
            const view = render(<Movies />);

            expect(view.container).toMatchSnapshot();
        });
    });

    describe('when movies is empty ', () => {
        beforeEach(() => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockUseSelector.mockImplementation((selectorFn: any) =>
                selectorFn({
                    movies: {
                        movies: [],
                    },
                }),
            );
        });

        it('should render correctly', () => {
            const view = render(<Movies />);

            expect(view.container).toMatchSnapshot();
        });
    });

    describe('when movies exist and is not empty', () => {
        beforeEach(() => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            mockUseSelector.mockImplementation((selectorFn: any) =>
                selectorFn({
                    movies: {
                        movies: mockMovies,
                    },
                }),
            );
        });

        it('should render correctly', () => {
            const view = render(<Movies />);

            expect(view.container).toMatchSnapshot();
        });
    });
});
