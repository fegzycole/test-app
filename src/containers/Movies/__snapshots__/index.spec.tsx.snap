// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Movies /> when movies exist and is not empty should render correctly 1`] = `
<div>
  <div
    class="MuiGrid-root css-1nx1jcs-MuiGrid-root-Container-wrapper"
  >
    <div
      class="MuiBox-root css-1pj0buh-StyledBox-wrapper"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1mj6qlk-MuiFormControl-root-MuiTextField-root-StyledTextField-wrapper"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r2:"
            placeholder="Enter Text Here"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-hdw1oc"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-1qhs0lv-MuiButtonBase-root-MuiIconButton-root-StyledButton-root"
        data-testid="btn"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-vr54bh-MuiSvgIcon-root-StyledSearchIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-1yk6gh-MuiGrid-root-StyledGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-1557hwu-MuiGrid-root-StytledGrid-wrapper"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-qegifv-MuiPaper-root-MuiCard-root-StyledCard-wrapper"
        >
          <img
            alt="movie-poster"
            class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            height="250"
            src="Mock Poster"
          />
          <div
            class="MuiCardContent-root css-1189jzg-MuiCardContent-root-StyledCardContent-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1qhct6m-MuiTypography-root-StyledButton-root"
            >
              Mock Title
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Movies /> when movies is empty  should render correctly 1`] = `
<div>
  <div
    class="MuiGrid-root css-1nx1jcs-MuiGrid-root-Container-wrapper"
  >
    <div
      class="MuiBox-root css-1pj0buh-StyledBox-wrapper"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1mj6qlk-MuiFormControl-root-MuiTextField-root-StyledTextField-wrapper"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            placeholder="Enter Text Here"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-hdw1oc"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-1qhs0lv-MuiButtonBase-root-MuiIconButton-root-StyledButton-root"
        data-testid="btn"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-vr54bh-MuiSvgIcon-root-StyledSearchIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-1yk6gh-MuiGrid-root-StyledGrid-root"
    >
      <div
        class="MuiBox-root css-qqlyqx-StyledBox-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          No Movie(s) with the title "" found
        </h5>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Movies /> when movies is undefined should render correctly 1`] = `
<div>
  <div
    class="MuiGrid-root css-1nx1jcs-MuiGrid-root-Container-wrapper"
  >
    <div
      class="MuiBox-root css-1pj0buh-StyledBox-wrapper"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1mj6qlk-MuiFormControl-root-MuiTextField-root-StyledTextField-wrapper"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            placeholder="Enter Text Here"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-hdw1oc"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-1qhs0lv-MuiButtonBase-root-MuiIconButton-root-StyledButton-root"
        data-testid="btn"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-vr54bh-MuiSvgIcon-root-StyledSearchIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-1yk6gh-MuiGrid-root-StyledGrid-root"
    >
      <div
        class="MuiBox-root css-qqlyqx-StyledBox-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Welcome to Our service, please type the title of the movie you would like to search for in the input field above!
        </h5>
      </div>
    </div>
  </div>
</div>
`;
